### ISSUE TITLE
[CI] ParentIdAggIT.testParentIdAggregation failed

### ISSUE THREAD

**Issue by cbuescher** [elastic/elasticsearch#33752]
Build: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+periodic/2/console

This does not reproduced locally for me so far:
```
./gradlew :server:integTest \
  -Dtests.seed=4AF65CDEFFB3B80B \
  -Dtests.class=org.elasticsearch.search.aggregations.bucket.ParentIdAggIT \
  -Dtests.method="testParentIdAggregation" \
  -Dtests.security.manager=true \
  -Dtests.locale=pt-PT \
  -Dtests.timezone=America/Blanc-Sablon \
  -Dcompiler.java=10 \
  -Druntime.java=8
```

Failure:
```
23:34:09 ERROR   0.47s J0 | ParentIdAggIT.testParentIdAggregation <<< FAILURES!
23:34:09    > Throwable #1: java.util.NoSuchElementException
23:34:09    > 	at __randomizedtesting.SeedInfo.seed([4AF65CDEFFB3B80B:244B697A07093F1F]:0)
23:34:09    > 	at java.util.AbstractList$Itr.next(AbstractList.java:364)
23:34:09    > 	at org.elasticsearch.search.aggregations.bucket.ParentIdAggIT.testParentIdAggregation(ParentIdAggIT.java:56)
23:34:09    > 	at java.lang.Thread.run(Thread.java:748)
```

Later on there are connection exceptions in the log, but I guess they could be due to an unclean shutdown:
```
23:34:09   1> [2018-09-14T17:34:09,340][WARN ][o.e.c.NodeConnectionsService] [node_sc3] failed to connect to node {node_s1}{WRWl6YZtSDaWqY8uMTNkkw}{f2pm5eBYQja1XF7fFoDRaw}{127.0.0.1}{127.0.0.1:36048} (tried [1] times)
23:34:09   1> org.elasticsearch.transport.ConnectTransportException: [node_s1][127.0.0.1:36048] connect_exception
23:34:09   1> 	at org.elasticsearch.transport.TcpChannel.awaitConnected(TcpChannel.java:165) ~[main/:?]
23:34:09   1> 	at org.elasticsearch.transport.TcpTransport.openConnection(TcpTransport.java:454) ~[main/:?]
23:34:09   1> 	at org.elasticsearch.transport.TcpTransport.openConnection(TcpTransport.java:117) ~[main/:?]
23:34:09   1> 	at org.elasticsearch.transport.ConnectionManager.internalOpenConnection(ConnectionManager.java:238) ~[main/:?]
23:34:09   1> 	at org.elasticsearch.transport.ConnectionManager.connectToNode(ConnectionManager.java:120) ~[main/:?]
23:34:09   1> 	at org.elasticsearch.transport.TransportService.connectToNode(TransportService.java:369) ~[main/:?]
23:34:09   1> 	at org.elasticsearch.transport.TransportService.connectToNode(TransportService.java:356) ~[main/:?]
23:34:09   1> 	at org.elasticsearch.cluster.NodeConnectionsService.validateAndConnectIfNeeded(NodeConnectionsService.java:153) [main/:?]
23:34:09   1> 	at org.elasticsearch.cluster.NodeConnectionsService$1.doRun(NodeConnectionsService.java:106) [main/:?]
23:34:09   1> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:723) [main/:?]
23:34:09   1> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [main/:?]
23:34:09   1> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_181]
23:34:09   1> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_181]
23:34:09   1> 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_181]
23:34:09   1> Caused by: java.net.ConnectException: Connection refused
23:34:09   1> 	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) ~[?:?]
23:34:09   1> 	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:717) ~[?:?]
23:34:09   1> 	at org.elasticsearch.transport.nio.channel.NioSocketChannel.internalFinish(NioSocketChannel.java:174) ~[framework-6.5.0-SNAPSHOT.jar:6.5.0-SNAPSHOT]
23:34:09   1> 	at org.elasticsearch.transport.nio.channel.NioSocketChannel.finishConnect(NioSocketChannel.java:152) ~[framework-6.5.0-SNAPSHOT.jar:6.5.0-SNAPSHOT]
23:34:09   1> 	at org.elasticsearch.transport.nio.SocketSelector.attemptConnect(SocketSelector.java:221) ~[framework-6.5.0-SNAPSHOT.jar:6.5.0-SNAPSHOT]
23:34:09   1> 	at org.elasticsearch.transport.nio.SocketSelector.setupChannel(SocketSelector.java:210) ~[framework-6.5.0-SNAPSHOT.jar:6.5.0-SNAPSHOT]
23:34:09   1> 	at org.elasticsearch.transport.nio.SocketSelector.setUpNewChannels(SocketSelector.java:198) ~[framework-6.5.0-SNAPSHOT.jar:6.5.0-SNAPSHOT]
23:34:09   1> 	at org.elasticsearch.transport.nio.SocketSelector.preSelect(SocketSelector.java:75) ~[framework-6.5.0-SNAPSHOT.jar:6.5.0-SNAPSHOT]
23:34:09   1> 	at org.elasticsearch.transport.nio.ESSelector.singleLoop(ESSelector.java:102) ~[framework-6.5.0-SNAPSHOT.jar:6.5.0-SNAPSHOT]
23:34:09   1> 	at org.elasticsearch.transport.nio.ESSelector.runLoop(ESSelector.java:78) ~[framework-6.5.0-SNAPSHOT.jar:6.5.0-SNAPSHOT]
23:34:09   1> 	... 1 more
```

**Comment by cbuescher** [comment]
Another one here: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-darwin-compatibility/2/console

**Comment by atorok** [comment]
and another https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+matrix-java-periodic/ES_BUILD_JAVA=java10,ES_RUNTIME_JAVA=java8fips,nodes=virtual&&linux/10/console

**Comment by atorok** [comment]
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+periodic/27/console

**Comment by romseygeek** [comment]
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-unix-compatibility/os=debian/13/console

Muting this now

**Comment by cbuescher** [comment]
I was able to see the same failure locally now a couple of times, although it rarely occurs if running the test separately. I had to run with "-Dtests.iters" (5 usually seems to trigger at least one failure, sometimes more). Whats odd is the test expects two child documents in the aggregation back (terms aggregation on the `_parent#parenttype` field, and the search indeed returns two documents but the agg is empty. This is the search response in a failure case:

```
{
	"took": 4,
	"timed_out": false,
	"_shards": {
		"total": 4,
		"successful": 4,
		"skipped": 0,
		"failed": 0
	},
	"_clusters": {
		"total": 0,
		"successful": 0,
		"skipped": 0
	},
	"hits": {
		"total": 2,
		"max_score": 1.0,
		"hits": [{
			"_index": "testidx",
			"_type": "childtype",
			"_id": "XbW9IGYBDXToQi7_snDQ",
			"_score": 1.0,
			"_routing": "p1",
			"_source": {
				"num": 1
			}
		}, {
			"_index": "testidx",
			"_type": "childtype",
			"_id": "XrW9IGYBDXToQi7_snD0",
			"_score": 1.0,
			"_routing": "p1",
			"_source": {
				"num": 2
			}
		}]
	},
	"aggregations": {
		"children": {
			"doc_count_error_upper_bound": 0,
			"sum_other_doc_count": 0,
			"buckets": []
		}
	}
}
```

I first thought the index might not be ready due to some timing issues but this seems to be okay.
The reason why this failed only on 6.x (approximately since Sept. 14th) is because the test doesn't exist any more on master.

**Comment by iverase** [comment]
@cbuescher I had a look to this test and I could reproduce it as well locally. What I notice that the errors always happen when you insert the data and the cluster is yellow:

```
[2018-09-28T17:05:58,074][INFO ][o.e.c.m.MetaDataCreateIndexService] [node_sm0] [testidx] creating index, cause [api], templates [random_index_template], shards [2]/[1], mappings [childtype]
[2018-09-28T17:05:58,183][INFO ][o.e.c.m.MetaDataMappingService] [node_sm0] [testidx/6GPfiwfhR4yJji67XO_1fA] update_mapping [childtype]
[2018-09-28T17:05:58,293][INFO ][o.e.c.r.a.AllocationService] [node_sm0] Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[testidx][1]] ...]).
```
instead of:

```
[2018-09-28T17:06:01,197][INFO ][o.e.c.m.MetaDataCreateIndexService] [node_sm0] [testidx] creating index, cause [api], templates [random_index_template], shards [5]/[0], mappings [childtype]
[2018-09-28T17:06:01,258][INFO ][o.e.c.r.a.AllocationService] [node_sm0] Cluster health status changed from [YELLOW] to [GREEN] (reason: [shards started [[testidx][2], [testidx][3], [testidx][0]] ...]).
[2018-09-28T17:06:01,274][INFO ][o.e.c.m.MetaDataMappingService] [node_sm0] [testidx/2gBOHkzPSWimxQt4keeLag] update_mapping [childtype]
```
 
If you add an `ensureGreen` statement before inserting the data then the errors go away. But I think that is actually hiding the real problem.

hope this helps!

**Comment by iverase** [comment]
I performed another check. If I wait one second before executing the query it seems that the errors goes away as well. My suspicion is that a shard gets searchable before all the info of parent-child type gets copied over. Therefore you find the docs but the aggregation fails.

**Comment by cbuescher** [comment]
>  If I wait one second before executing the query it seems that the errors goes away

I also tried this (even 2 sec) but it doesn't work for me at lease. Still random errors.

> My suspicion is that a shard gets searchable before all the info of parent-child type gets copied over

This is what I already suspect, but still trying to confirm.

**Comment by cbuescher** [comment]
This should be solved now that #34259 is merged. Closing.
